NFT Tutorial readme.md (#1)
tqtezos/nft-tutorial: README.md

implementation. The tutorial uses pre-compiled FA2 NFT contract written in
[LIGO](https://ligolang.org/) smart contract language and command line interface
(CLI) to originate and interact with the NFT contracts either on the
[Flextesa sandbox](https://tezos.gitlab.io/flextesa/) or Tezos testnet (Carthagenet).

## Introduction

### What is FA2

FA2 refers to a token standard ([TZIP-12](https://gitlab.com/tzip/tzip/-/blob/master/proposals/tzip-12/tzip-12.md))
on Tezos. FA2 proposes a unified token contract interface, supporting a wide range
of token types. The FA2 standard provides a standard API to transfer tokens, check
token balances,manage operators (an address that is permitted to manage tokens
on behalf of the token owner) and manage token metadata.

### What is NFT

NFT (non-fungible token) is a special type of cryptographic token which represents
something unique; non-fungible tokens are thus not mutually interchangeable.
NFTs can represent ownership over digital or physical assets like virtual collectibles
or unique art work.

For each individual non-fungible token the FA2 implementation assigns a unique

> token ID and associates it with the token owner address. The FA2 API allows to

Reviewer: allows to inspect -> enables the inspection of

inspect token balances for the specific token ID and token owner address. For NFTs
the balance can be either 0 (which means that the address does not own this particular
token) or 1 (the address owns the token).

The FA2 contract also associates some metadata with each token. This tutorial supports
token symbol and token name metadata attributes. However, the implementation can

> be easily extended to support custom metadata attributes such an associate image

Reviewer: associate -> associated

# FA2 NFT Tutorial

This tutorial shows how to originate and interact with the FA2 NFT contract
implementation. The tutorial uses a pre-compiled FA2 NFT contract written in
[LIGO](https://ligolang.org/) smart contract language and a command line interface
(CLI) to originate and interact with the NFT contracts either on the
[Flextesa](https://tezos.gitlab.io/flextesa/) sandbox or Tezos testnet (Carthagenet).

## Introduction

### What is FA2

FA2 refers to a token standard ([TZIP-12](https://gitlab.com/tzip/tzip/-/blob/master/proposals/tzip-12/tzip-12.md))
on Tezos. FA2 proposes a unified token contract interface, supporting a wide range
of token types. The FA2 provides a standard API to transfer tokens, check
token balances, manage operators (addresses that are permitted to transfer tokens
on behalf of the token owner) and manage token metadata.

### What is NFT

NFT (non-fungible token) is a special type of cryptographic token which represents
something unique; non-fungible tokens are thus not mutually interchangeable.
NFTs can represent ownership over digital or physical assets like virtual collectibles

> or unique art work.

Reviewer 2: art work -> artwork

token symbol and token name metadata attributes. However, the implementation can
be easily extended to support custom metadata attributes such an associated image
or document URL and its crypto-hash.

## Tutorial

### Prerequisites

- [Node.js](https://nodejs.org/) must be installed. The Node installation must also
  include `npm` (Node package manager).

- [Docker](https://www.docker.com/) must be installed. You need docker to run
  Flextesa sandbox. You might skip docker installation if you plan to run this
  tutorial on the testnet (Carthagenet) only.

### The CLI Tool

You will need to install `tznft` CLI tool. After the installation, you can invoke
various commands in the form of `tznft <command> [options]`. `tznft` provides the
following commands:

- NFT mint (contract origination) and token inspection command
- NFT transfer command
- Configuration commands to bootstrap Tezos network and configure address aliases

The commands will be explained in more detail below. You can always run

```sh
$ tznft --help
```

to list all available commands.

### Initial Setup

1. Create a new local directory to keep your tutorial configuration:

> 

Reviewer 2: do we really need a directory to keep the config file? and if so, why not just create it for the user, e.g. at ~/.nft-tutorial/tznft.json

Reviewer 2: also, if tznft.json is in a specific location rather than current directory then tznft, being globally installed, can be successfully executed regardless of current directory, whereas currently it prints error `unknown error:
Error: no tznft.json config file found` if executed in the "wrong" place. Also, missing config file appears to be a pretty well known error and handled as such, so stack trace shouldn't be printed

Author: The idea is that user can edit it manually and add additional networks for example.

Reviewer 2: they can edit it at a well-known location ~/.nft-tutorial/tznft.json just as well

   ```sh
   $ mkdir nft-tutorial
   $ cd nft-tutorial
   ```

2. Install `@tqtezos/nft-tutorial` npm package:

   ```sh
   $ npm install -g https://github.com/tqtezos/nft-tutorial.git

   /usr/local/bin/tznft -> /usr/local/lib/node_modules/@tqtezos/nft-tutorial/lib/tznft.js

   + @tqtezos/nft-tutorial@1.0.0
   added 3 packages from 1 contributor and updated 145 packages in 11.538s
   ```

   The command installs `tznft` CLI tool.

3. Initialize tutorial config:

   ```sh
   $ tznft config-init

>    tznft.json config file created

Reviewer 2: would be nice to print full, absolute path here (especially if config file is created not in the current directory as per my suggestion above)

   ```

4. Check that the default active network is `sandbox`:

   ```sh
   $ tznft show-network

   active network: sandbox
   ```

5. Bootstrap Tezos network:

   ```sh
   $ tznft bootstrap

   ebb03733415c6a8f6813a7b67905a448556e290335c5824ca567badc32757cf4

   starting sandbox...
   sandbox started
   originating balance inspector contract...
   originated balance inspector KT1Pezr7JjgmrPcPhpkbkH1ytG7saMZ34sfd
   ```

   If you are bootstrapping a `sandbox` network for the first time, Docker will download
   the Flextesa docker-image as well.

   The default configuration comes with two account aliases `bob` and `alice`
   that can be used for token minting and transferring.

### Mint NFT Token(s)

> This tutorial uses NFT collection contract. Each time the user mints a new set

Reviewer: uses -> uses an

updating operators...
updated operators
```

Now `bob` can transfer a token on behalf of `alice` again:

```sh
$ tznft transfer --nft KT1XP3RE6S9t44fKR9Uo5rAfqHvHXu9Cy7fh --signer bob --batch 'alice, bob, 1'

transferring tokens...
tokens transferred
```

Inspecting balances after the transfer:

```sh
$ tznft show-balance --nft KT1XP3RE6S9t44fKR9Uo5rAfqHvHXu9Cy7fh --signer bob --owner bob --tokens 0 1

querying NFT contract KT1XP3RE6S9t44fKR9Uo5rAfqHvHXu9Cy7fh using balance inspector KT1Pezr7JjgmrPcPhpkbkH1ytG7saMZ34sfd
requested NFT balances:
owner: tz1YPSCGWXwBdTncK2aCctSZAXWvGsGwVJqU	token: 0	balance: 0
owner: tz1YPSCGWXwBdTncK2aCctSZAXWvGsGwVJqU	token: 1	balance: 1

$ tznft show-balance --nft KT1XP3RE6S9t44fKR9Uo5rAfqHvHXu9Cy7fh --signer bob --owner alice --tokens 0 1

querying NFT contract KT1XP3RE6S9t44fKR9Uo5rAfqHvHXu9Cy7fh using balance inspector KT1Pezr7JjgmrPcPhpkbkH1ytG7saMZ34sfd
requested NFT balances:
owner: tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb	token: 0	balance: 1
owner: tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb	token: 1	balance: 0
```

Token `1` now belongs to `bob`.

### Configuration

`tznft` can be configured to interact with different Tezos networks. The user can

> also configure address aliases to sign Tezos operation and/or use them as command

Reviewer: Tezos operation -> Tezos operations

# FA2 NFT Tutorial

This tutorial shows how to originate and interact with the FA2 NFT contract
implementation. The tutorial uses pre-compiled FA2 NFT contract written in
[LIGO](https://ligolang.org/) smart contract language and command line interface
(CLI) to originate and interact with the NFT contracts either on the
[Flextesa](https://tezos.gitlab.io/flextesa/) sandbox or Tezos testnet (Carthagenet).

## Introduction

### What is FA2

FA2 refers to a token standard ([TZIP-12](https://gitlab.com/tzip/tzip/-/blob/master/proposals/tzip-12/tzip-12.md))
on Tezos. FA2 proposes a unified token contract interface, supporting a wide range
of token types. The FA2 provides a standard API to transfer tokens, check
token balances, manage operators (addresses that are permitted to transfer tokens
on behalf of the token owner) and manage token metadata.

### What is NFT

> NFT (non-fungible token) is a special type of cryptographic token which represents

Author: Copied definition from Wikipedia

> NFT (non-fungible token) is a special type of cryptographic token which represents

Reviewer: Maybe "blockchain asset?"

> NFT (non-fungible token) is a special type of cryptographic token which represents

Author: I do not have a strong opinion on this one

be easily extended to support custom metadata attributes such an associated image
or document URL and its crypto-hash.

## Tutorial

### Prerequisites

- [Node.js](https://nodejs.org/) must be installed. Node installation must also
  include `npm` (Node package manager).

- [Docker](https://www.docker.com/) must be installed. You need docker to run
  Flextesa sandbox. You might skip docker installation if you plan to run this
  tutorial on the testnet (Carthagenet) only.

### The CLI Tool

You will need to install `tznft` CLI tool. After the installation, you can invoke
various commands in the form of `tznft <command> [options]`. `tznft` provides the
following commands:

- NFT mint (contract origination) and token inspection command
- NFT transfer command
- Configuration commands to bootstrap Tezos network and configure address aliases

The commands will be explained in more details below. You can always run

```sh
$ tznft --help
```

to list all available commands.

### Initial Setup

1. Create a new local directory to keep your tutorial configuration:

   ```sh
   $ mkdir nft-tutorial
   $ cd nft-tutorial

>    ```

Author: Not on the repo READM.md. I will definitely look into it once I start migration to the assets web portal

   ```

   If you are bootstrapping `sandbox` network first time, Docker will download
   Flextesa sandbox image as well.

   The default configuration comes with two bootstrap aliases `bob` and `alice`
   that can be used for token minting and transferring.

### Mint NFT Token(s)

This tutorial uses NFT collection contract. Each time the user mints a new set
(collection) of tokens, a new NFT contract is created. The user cannot add more
tokens or remove (burn) existing tokens within the contract. However tokens can
be transferred to other owners.

`mint` command requires the following parameters:

- <owner> alias or address of the new tokens owner
- `--tokens` new tokens metadata. Each token metadata is represented as comma
  delimited string: `'<token_id>, <token_symbol>, <token_name>'`:

```sh
$ tznft mint <owner_alias> --tokens <token_meta_list>`
```

Example:

```sh
$ tznft mint bob --tokens '0, T1, My Token One' '1, T2, My Token Two'

originating new NFT contract...
originated NFT collection KT1XP3RE6S9t44fKR9Uo5rAfqHvHXu9Cy7fh
```

### Inspecting The NFT Contract

Using `KT1..` address of the NFT contract created by the `mint` command, we can

> inspect token metadata and balances (i. e. what addresses own the tokens).

Reviewer 3: ```suggestion
inspect token metadata and balances (i. e. which addresses own the tokens).
```

#### Inspect Token Metadata

`show-meta` command requires the following parameters:

- `--nft` address of the FA2 NFT contract to inspect
- `--operator` alias on behalf of which contract is inspected
- `--tokens` a list of token IDs to inspect

```sh
$ tznft show-meta --nft <nft_address> --operator <alias> --tokens <token_id_list>
```

Example:

```sh
$ tznft show-meta --nft KT1XP3RE6S9t44fKR9Uo5rAfqHvHXu9Cy7fh --operator bob --tokens 0 1

token_id: 0	symbol: T1	name: My Token One	extras: { }
token_id: 1	symbol: T2	name: My Token Two	extras: { }
```

#### Inspect Token Balances

`show-balance` command requires the following parameters:

- `--nft` address of the FA2 NFT contract to inspect

> - `--operator` alias on behalf of which contract is inspected

Reviewer 3: It's confusing to use the word operator here. It's not really an FA2-operator *and* the account is only used to sign a dry-run operation (could be completely hidden from the user).

Author: it is not a dry run, it is a real run (taquito seems does not support dry-run).
The notion of the signing operator is required for `transfer` operation to demonstrate FA2 operators functionality.
So, all other commands that require signing also use `--operator` option.
However, I agree that the name "operator" might be confusing.
I am open for suggestions what to use instead. Maybe "signer"?

Reviewer 3: yes, signer works, there is also `--gas-account` that is used often.

```

Example, `bob` transfers his own tokens `0` and `1` to `alice`:

```sh
$ tznft transfer --nft KT1XP3RE6S9t44fKR9Uo5rAfqHvHXu9Cy7fh --operator bob --batch 'bob, alice, 0' 'bob, alice, 1'

transferring tokens...
tokens transferred
```

Now, we can check token balances after the transfer:

```sh
$ tznft show-balance --nft KT1XP3RE6S9t44fKR9Uo5rAfqHvHXu9Cy7fh --operator bob --owner bob --tokens 0 1

querying NFT contract KT1XP3RE6S9t44fKR9Uo5rAfqHvHXu9Cy7fh using balance inspector KT1Pezr7JjgmrPcPhpkbkH1ytG7saMZ34sfd
requested NFT balances:
owner: tz1YPSCGWXwBdTncK2aCctSZAXWvGsGwVJqU	token: 0	balance: 0
owner: tz1YPSCGWXwBdTncK2aCctSZAXWvGsGwVJqU	token: 1	balance: 0

$ tznft show-balance --nft KT1XP3RE6S9t44fKR9Uo5rAfqHvHXu9Cy7fh --operator bob --owner alice --tokens 0 1

querying NFT contract KT1XP3RE6S9t44fKR9Uo5rAfqHvHXu9Cy7fh using balance inspector KT1Pezr7JjgmrPcPhpkbkH1ytG7saMZ34sfd
requested NFT balances:
owner: tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb	token: 0	balance: 1
owner: tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb	token: 1	balance: 1
```

### Operator Transfer

It is also possible to transfer tokens on behalf of the owner.

> `bob` is trying to transfer one of the `alice` tokens back:

Reviewer 3: ```suggestion
`bob` is trying to transfer one of `alice`'s tokens back:
```

  Example:

  ```sh
  $ tznft show-network --all

  * sandbox
    testnet
  ```

- `bootstrap` bootstrap selected network and deploy helper balance inspector contract.
  If selected network is `sandbox` this command needs to be run each time sandbox
  is restarted, for other public networks like `testnet` is it enough to run this
  command once.

  Example:

  ```sh
  $ tznft bootstrap

  366b9f3ead158a086e8c397d542b2a2f81111a119f3bd6ddbf36574b325f1f03

  starting sandbox...
  sandbox started
  originating balance inspector contract...
  originated balance inspector KT1WDqPuRFMm2HwDRBotGmnWdkWm1WyG4TYE
  ```

- `kill-sandbox` stop Flextesa sandbox process if selected network is `sandbox`.
  This command has no effect on other network types.

  Example:

  ```sh
  $ tznft kill-sandbox

  flextesa-sandbox

  killed sandbox.
  ```

The sandbox network (selected by default) is configured to bake new Tezos blocks
every 5 seconds. It makes running the commands that interact with the network
faster. However, all originated contracts will be lost after the sandbox is stopped.

If you are using `testnet`, your originated contracts will remain on the blockchain

> and you can inspect them afterwards using an indexer like [BCD](https://better-call.dev/).

Reviewer 3: ```suggestion
and you can inspect them afterwards using an block explorer like [BCD](https://better-call.dev/).
```

#### Alias Configuration Commands

`tznft` allows user to configure and use short names (aliases) instead of typing
in full Tezos addresses when invoking `tznft` commands.
Each network comes with two pre-configured aliases `bob` and `alice`. The user
can manage aliases by directly editing `tznft.json` file or using the following
commands:

- `show-alias [alias]` show address and private key (if configured) of the
  specified `[alias]`. If `[alias]` option is not specified, show all configured
  aliases.

  Example:

  ```sh
  $ tznft show-alias bob

  bob	tz1YPSCGWXwBdTncK2aCctSZAXWvGsGwVJqU	edsk3RFgDiCt7tWB2oe96w1eRw72iYiiqZPLu9nnEY23MYRp2d8Kkx

  $ tznft show-alias

  bob	tz1YPSCGWXwBdTncK2aCctSZAXWvGsGwVJqU	edsk3RFgDiCt7tWB2oe96w1eRw72iYiiqZPLu9nnEY23MYRp2d8Kkx
  alice	tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb	edsk3QoqBuvdamxouPhin7swCvkQNgq4jP5KZPbwWNnwdZpSpJiEbq
  ```

> - `add-alias <alias> <private_key>` add alias using its private key. Aliases

Reviewer 3: Does it says somewhere what kind of secret-keys are accepted? (encrypted, tz2, tz3, etc.)

Author: Whatever keys are supported by Taquito, I believe it supports all tz1 tz2, tz3 types of keys

  that configured with the private key can be used to sign operations that
  originate or call smart contracts on chain. `tznft` commands that require Tezos
  operation signing have `--operator` option (not to be confused with the FA2
  token operator).

  Example:

  ```sh
  $ tznft add-alias jane edsk3QoqBuvdamxouPhin7swCvkQNgq4jP5KZPbwWNnwdZpSpJiEbq

  alias jane has been added

  $ tznft show-alias jane

  jane	tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb	edsk3QoqBuvdamxouPhin7swCvkQNgq4jP5KZPbwWNnwdZpSpJiEbq
  ```

- `add-alias <alias> <address>` add alias using Tezos address (public key hash).
  Such aliases do not have associated private key and cannot be used to sign
  Tezos operations.

  Example:

  ```sh
  $ tznft add-alias michael tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb

  alias michael has been added

  $ tznft show-alias michael

  michael	tz1VSUr8wwNhLAzempoch5d6hLRiTh8Cjcjb
  ```

- `add-alias-faucet <alias> <faucet_json_file_path>` add alias with private key
  from the faucet file (see [Tezos Faucet](https://faucet.tzalpha.net/)). This

>   command will not work on `sandbox` network. Alias configured from the faucet

Reviewer 3: ```suggestion
  command will not work on `sandbox` network. An alias configured from the faucet
```
